File: day7/ans.py
from collections import deque

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        
        queue = deque()
        queue.append((root, 0, 0))
        
        x_depth = None
        y_depth = None
        x_parent_val = None
        y_parent_val = None
        
        while queue:
            node, depth, parent_val = queue.popleft()
            
            if node.val == x:
                x_depth = depth
                x_parent_val = parent_val
            
            if node.val == y:
                y_depth = depth
                y_parent_val = parent_val         
            
            if node.left:
                queue.append((node.left, depth + 1, node.val))
            
            if node.right:
                queue.append((node.right, depth + 1, node.val))
        
        if x_depth != None and x_depth == y_depth and x_parent_val != y_parent_val:
            return True
        
        return False

File: day7/ans.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        queue<TreeNode*> q;
        q.push(root);
        
        while(!q.empty()) {
            //Traversing the level
            auto level_size = q.size();
            bool x_exists = false;
            bool y_exists = false;
            
            for(int i = 0; i < level_size; i++){
                auto ele = q.front();
                
                if(ele->val == x) {
                    x_exists = true;
                }
                
                if(ele->val == y) {
                    y_exists = true;
                }
                
                //making sure that they are not siblings -> same parent for next level
                if(ele->left && ele->right) {
                    if(ele->left->val == x && ele->right->val == y) {
                        return false;
                    }
                    
                    if(ele->left->val == y && ele->right->val == x) {
                        return false;
                    }
                }
                
                if(ele->left) {
                    q.push(ele->left);
                }
                
                if(ele->right) {
                    q.push(ele->right);
                }
                
                q.pop();
                
            }
            
            //if both exists, they are cousins. Siblings are ruled out in previous level traversal
            //if only one exists, then other one is present in some other level, thus return false
            if(x_exists && y_exists) {
                return true;
            } else if(x_exists || y_exists) {
                return false;
            }
        }
        
        return false;
    }
};
